<a href="https://colab.research.google.com/github/Fata-Haidar/244107020108-Pembelajaran-Mesin/blob/main/JS05/JS05_tugas2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Import Library**

In [1]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import (
    CountVectorizer, TfidfVectorizer
)
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import (
    accuracy_score,
    classification_report,
    confusion_matrix
)

**Load Data**

In [2]:
# Membaca dataset
df = pd.read_csv('spam.csv', encoding='latin-1')

# Menghapus 3 kolom yang tidak digunakan
df = df.drop(df.iloc[:, 2:], axis=1)

# Mengubah nama kolom
df.columns = ['Labels', 'SMS']

# Cek data
df.head()

,Labels,SMS
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


**Preprocessing**

In [3]:
# Encode label
new_labels = {
    'spam': 1,
    'ham': 0
}

df['Labels'] = df['Labels'].map(new_labels)

# Memisahkan fitur dan label
X = df['SMS'].values
y = df['Labels'].values

# Membagi data
X_train_text, X_test_text, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=50
)

Data dibagi menggunakan konfigurasi yang sama seperti percobaan pada lab 3 sebelumnya supaya hasilnya bisa dibandingkan.

**1. Klasifikasi Multinomial Naive Bayes menggunakan CountVectorizer dengan stop_words**

In [4]:
# Inisiasi CountVectorizer dengan stop_words
bow = CountVectorizer(stop_words='english')

# Transformasi data
X_train_bow = bow.fit_transform(X_train_text)
X_test_bow = bow.transform(X_test_text)

# Membuat model Multinomial Naive Bayes
mnb_bow = MultinomialNB()

# Melatih model
mnb_bow.fit(X_train_bow, y_train)

# Prediksi data training dan testing
y_pred_train_bow = mnb_bow.predict(X_train_bow)
y_pred_test_bow = mnb_bow.predict(X_test_bow)

# Evaluasi model
acc_train_bow = accuracy_score(y_train, y_pred_train_bow)
acc_test_bow = accuracy_score(y_test, y_pred_test_bow)

print(f'Akurasi training: {acc_train_bow:.4%}')
print(f'Akurasi testing: {acc_test_bow:.4%}')

print('\nClassification Report:')
print(classification_report(
    y_test,
    y_pred_test_bow,
    target_names=['ham', 'spam']
))

print('\nConfusion Matrix:')
print(confusion_matrix(y_test, y_pred_test_bow))

Akurasi training: 99.4615%
Akurasi testing: 98.2960%

Classification Report:
              precision    recall  f1-score   support

         ham       0.98      1.00      0.99       954
        spam       0.98      0.90      0.94       161

    accuracy                           0.98      1115
   macro avg       0.98      0.95      0.96      1115
weighted avg       0.98      0.98      0.98      1115


Confusion Matrix:
[[951   3]
 [ 16 145]]


Pada percobaan ini, saya menggunakan CountVectorizer dengan mengaktifkan stop_words='english' untuk menghilangkan kata-kata umum yang kurang memberikan informasi dalam klasifikasi SMS. Setelah itu, saya melatih model Multinomial Naive Bayes dan memperoleh akurasi testing sebesar 98.296%.

**2. Klasifikasi Multinomial Naive Bayes menggunakan TF-IDF dengan stop_words**

In [5]:
# Inisiasi TF-IDF dengan stop_words
tfidf = TfidfVectorizer(stop_words='english')

# Transformasi data
X_train_tfidf = tfidf.fit_transform(X_train_text)
X_test_tfidf = tfidf.transform(X_test_text)

# Membuat model Multinomial Naive Bayes
mnb_tfidf = MultinomialNB()

# Melatih model
mnb_tfidf.fit(X_train_tfidf, y_train)

# Prediksi data training dan testing
y_pred_train_tfidf = mnb_tfidf.predict(X_train_tfidf)
y_pred_test_tfidf = mnb_tfidf.predict(X_test_tfidf)

# Evaluasi model
acc_train_tfidf = accuracy_score(y_train, y_pred_train_tfidf)
acc_test_tfidf = accuracy_score(y_test, y_pred_test_tfidf)

print(f'Akurasi training: {acc_train_tfidf:.4%}')
print(f'Akurasi testing: {acc_test_tfidf:.4%}')

print('\nClassification Report:')
print(classification_report(
    y_test,
    y_pred_test_tfidf,
    target_names=['ham', 'spam']
))

print('\nConfusion Matrix:')
print(confusion_matrix(y_test, y_pred_test_tfidf))

Akurasi training: 98.4294%
Akurasi testing: 96.0538%

Classification Report:
              precision    recall  f1-score   support

         ham       0.96      1.00      0.98       954
        spam       1.00      0.73      0.84       161

    accuracy                           0.96      1115
   macro avg       0.98      0.86      0.91      1115
weighted avg       0.96      0.96      0.96      1115


Confusion Matrix:
[[954   0]
 [ 44 117]]


Pada percobaan ini, saya menggunakan TF-IDF dengan mengaktifkan stop words untuk mengurangi kata-kata umum yang kurang penting. Setelah model dilatih dan diuji, saya memperoleh akurasi testing sebesar 96.0538%.

**Perbandingan hasil**

In [6]:
# Membandingkan akurasi
perbandingan = pd.DataFrame({
    'Metode': ['CountVectorizer', 'TF-IDF'],
    'Training': [acc_train_bow, acc_train_tfidf],
    'Testing': [acc_test_bow, acc_test_tfidf]
})

# Menampilkan dalam persentase
perbandingan[['Training', 'Testing']] *= 100

print(perbandingan.round(2))

            Metode  Training  Testing
0  CountVectorizer     99.46    98.30
1           TF-IDF     98.43    96.05


Berdasarkan percobaan, saya menyimpulkan bahwa CountVectorizer dengan stop words lebih baik karena menghasilkan akurasi 98,30%, sedangkan TF-IDF hanya 96,05%.